video: addition of __ASSERT to ensure proper arguments in api functions

avolmat-st opened the pull request:
Over the video api function, some functions perform some pointer validation, via if statements or via __ASSERT while some other don't.
Some drivers also performs some pointers validity check (mostly on fmt pointers) while other don't.

In order to try to have something similar for all and avoid drivers having to think about wherher pointer checking is necessary or not, perform those checks within all API functions. Within api ops, drivers can now consider that all pointers are valid.

Commit:
video: common: add __ASSERT_NO_MSG in video api function entry

Protect video API functions via __ASSERT_NO_MSG call to ensure that
required pointers are valid when entering functions.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: remove valid pointer check in api entry points

Now that video api functions have __ASSERT_NO_MSG calls, drivers
do not need to check the entry point functions pointers.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

ngphibang reviewed (dismissed):
LGTM. Thanks !

BTW, I am not sure between `__ASSERT` and `return -EINVAL`, which one is more appropriate ? Returning error codes can let the program continue but require the caller to check the return value, ASSERT is simpler (I am up for ASSERT) ...

avolmat-st commented:
> LGTM. Thanks !
> 
> BTW, I am not sure between `__ASSERT` and `return -EINVAL`, which one is more appropriate ? Returning error codes can let the program continue but require the caller to check the return value, ASSERT is simpler (I am up for ASSERT) ...

I am been wondering myself as well. Here I think all parameters checked with ASSERT are never expected to have such value, meaning that there is for sure an issue in the application or caller, thus it seemed to be that anyway it doesn't make sense to let the program proceed, hence using ASSERT.

pdgendt commented on drivers/video/video_common.c:
```diff
@@ -82,6 +82,8 @@ void video_buffer_release(struct video_buffer *vbuf)
 	struct mem_block *block = NULL;
 	int i;
 
+	__ASSERT(vbuf, "Invalid arguments");
```
Do not use implicit pointer to boolean conversion, this is a required rule, see https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html rule 85.

```suggestion
	__ASSERT(vbuf != NULL, "Invalid arguments");
```

pdgendt commented on drivers/video/video_common.c:
```diff
@@ -99,6 +101,8 @@ void video_buffer_release(struct video_buffer *vbuf)
 int video_format_caps_index(const struct video_format_cap *fmts, const struct video_format *fmt,
 			    size_t *idx)
 {
+	__ASSERT_NO_MSG(fmts != NULL && fmt != NULL && idx != NULL);
```
I didn't think of this last time I reviewed, but it is better to split these. Otherwise you don't know which assert failed.

```suggestion
	__ASSERT_NO_MSG(fmts != NULL);
	__ASSERT_NO_MSG(fmt != NULL);
	__ASSERT_NO_MSG(idx != NULL);
```

josuah reviewed (dismissed):
Thank you for the modification!

pdgendt commented on include/zephyr/drivers/video.h:
```diff
@@ -771,6 +816,8 @@ int video_format_caps_index(const struct video_format_cap *fmts, const struct vi
  */
 static inline uint64_t video_frmival_nsec(const struct video_frmival *frmival)
 {
+	__ASSERT_NO_MSG(frmival != NULL);
+
```
Nit, maybe add

```suggestion
	__ASSERT_NO_MSG(frmival != NULL);
	__ASSERT_NO_MSG(frmival->denominator != 0U);

```

josuah replied:
I remember of at least one bug I wrote (private code) that `frmival->denominator != 0U` could have avoided!

It could be worth to add `frmival->numerator != 0U` never 0, as 0 FPS is unlikely.

avolmat-st replied:
Yes. However I am wondering if this should be a ASSERT_NO_MSG in this case or just a return -EINVAL. Valid pointer checking and invalid values of frame interval is slightly different.
Once this is added in the common API code, the check done within the video_sw_generator can be thus removed.

avolmat-st replied:
Add some more __ASSERT_NO_MSG on some mandatory pointers within video api structs.
Add also check of numerator / denominator for frame interval and removed the check in sw_generator.

I haven't added check again some element of structs such as buf type, or frame interval type.

Commit:
video: common: add checks of frame interval values

Add checks regarding specific values (numerator / denominator) not being
0 for frame intervals.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: sw_generator: remove frmival values check

Now that api core functions check that denominator / numerator are not
null, it is no more necessary to perform this check within drivers.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
